alacritty_terminal/src/config/mod.rs

```diff
@@ -432,3 +413,97 @@ where
         },
     }
 }
+
+struct OrNoneVisitor<T>(PhantomData<fn() -> T>);
+
+impl<'de, T> Visitor<'de> for OrNoneVisitor<T>
+    where T: Deserialize<'de>
+{
+    type Value = Option<T>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("not null (use None instead)")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+        where E: de::Error
+    {
+        if value.to_lowercase() == "none" {
+            Ok(None)
+        } else {
+            Deserialize::deserialize(value.into_deserializer()).map(Some)
+        }
+    }
+
+    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
+        where E: de::Error
+    {
+        Deserialize::deserialize(value.into_deserializer()).map(Some)
+    }
+
+    fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
+        where E: de::Error
+    {
+        Deserialize::deserialize(value.into_deserializer()).map(Some)
+    }
+
+    fn visit_f64<E>(self, value: f64) -> Result<Self::Value, E>
+        where E: de::Error
+    {
+        Deserialize::deserialize(value.into_deserializer()).map(Some)
+    }
+
+    fn visit_bool<E>(self, value: bool) -> Result<Self::Value, E>
+        where E: de::Error
+    {
+        Deserialize::deserialize(value.into_deserializer()).map(Some)
+    }
+
+    fn visit_map<M>(self, map: M) -> Result<Self::Value, M::Error>
+        where M: MapAccess<'de>
+    {
+        Deserialize::deserialize(MapAccessDeserializer::new(map)).map(Some)
+    }
+
+    fn visit_seq<S>(self, seq: S) -> Result<Self::Value, S::Error>
+        where S: SeqAccess<'de>
+    {
+        Deserialize::deserialize(SeqAccessDeserializer::new(seq)).map(Some)
+    }
+}
+
+pub fn or_none<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
+    where D: Deserializer<'de>, T: Deserialize<'de>
+{
+    deserializer.deserialize_any(OrNoneVisitor(PhantomData))
+}
+
+struct OrStrVisitor<T>(PhantomData<fn() -> T>);
+
+impl<'de, T> Visitor<'de> for OrStrVisitor<T>
+    where T: Deserialize<'de> + FromStr<Err = Void>
+{
+    type Value = T;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("string or mapping")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+        where E: de::Error
+    {
+        Ok(Self::Value::from_str(value).ok().unwrap())
+    }
+
+    fn visit_map<M>(self, map: M) -> Result<Self::Value, M::Error>
+        where M: MapAccess<'de>
+    {
+        Deserialize::deserialize(MapAccessDeserializer::new(map))
+    }
+}
+
+pub fn struct_or_str<'de, T, D>(deserializer: D) -> Result<T, D::Error>
+where D: Deserializer<'de>, T: Deserialize<'de> + FromStr<Err = Void>
+{
+    deserializer.deserialize_any(OrStrVisitor(PhantomData))
+}
```

Comment:
This deserializer seems unnecessarily complex. Is there a reason why you did not follow my recommendation to just deserialize using a generic type parameter?

Reply:
I'm not super familiar with serde, but using a custom visitor seems to be the right way to 
intermediate the deserialization process.  This particular use is actually one of the documentation examples: https://serde.rs/string-or-struct.html

`serde_yaml::Value` doesn't implement `IntoDeserializer`, so in order to feed it back into deserialization after handling the special case I'd need to re-serialize it.  This avoids the intermediate data structure.  The same applies to the `or_none` helper.

Reply:
You can just deserialize it to Value and then use T::deserialize(value) to deserialize it from the value to the T, as long as T implements deserialize. The method shouldn't look much different than the working directory deserializer. You can look at failure_default as an example.

Reply:
It looks like that would work, but this method is more efficient since it doesn't create the intermediate object tree.

Reply:
It's also unmaintainable. The performance for the config file deserialization is not an issue. Maintaining the code for it is a lot more difficult though.